src/Mono.Android/Android.Runtime/JavaSet.cs

```diff
@@ -381,7 +381,7 @@ public void CopyTo (T[] array, int array_index)
 				throw new ArgumentException ("array");
 
 			int i = 0;
-			foreach (T item in this)
+			foreach (T item in this!)
```

Comment:
Any link to what this does?  I would *think* that `this!` asserts that `this` is not null, but that would be silly as `this` *can't* be null.  So does this do something different within the context of a `foreach` loop?  This confuses me.

Reply:
Okay… but why are we returning a "maybe-null" enumerator?

I "normally" see `[MaybeNull]` used when a return type is a generic type parameter, as (in C# 8) generic type parameters couldn't be null:

```csharp
// C#8
[return: MaybeNull]
static T GetValueOrDefault<T>(T value) => value ?? default;

// C#9 hotness
static T? GetValueOrDefault2<T>(T? value) => value ?? default;
```

I think this was needed at the time because the "obvious" thing of just using `T?` as the return type wasn't supported in C#8, but it appears to work with C#9…

Which brings me to my confusion part 1: why are methods which return `IEnumerable<T>` using `[return:MaybeNull]` in the first place?  C#8 would have supported `IEnumerable<T>?`, no?

Then there's confusion part 2: consider the `Extensions.ToEnumerable<T>()` and `Extensions.ToEnumerator_Dispose<T>()` methods: https://github.com/xamarin/xamarin-android/blob/941e04c22546227c5e1ae0eca9cda30740d14f9e/src/Mono.Android/System.Linq/Extensions.cs#L43-L65

Those are `yield return` methods.  They will *never* return a `null` value.  A value *in the sequence* could be `null`, e.g. the `JavaConvert.FromJniHandle(…)` call could return `null` , and that would be better expressed as:

```csharp
static IEnumerator<T?> ToEnumerator_Dispose<T>(this Java.Util.IIterator source) => …
```

The above works in C#9, but could not be expressed in C#8.  Fortunately we can now use C#9…

So unless I'm missing the mark somewhere, I think the "better" approach would be to:

1. Audit all usage of `[MaybeNull]` and see if it's still needed in a post-C#9 world.
2. Iterator methods will *never* return null, but their elements *could* be null, and as C#9 supports this, we should update our methods accordingly.  For example, `Extensions.ToEnumerator_Dispose<T>(IIterator)` should *drop* the `[return:MaybeNull]` and change the return type from `IEnumerator<T>` to `IEnumerator<T?>`.
3. (2) means that the `foreach` loop in `JavaSet<T>` which started this whole discussion would become:

    ```csharp
    foreach (T? item in this) { … }
    ```
    with no need for the `!` operator.
 4. (3) in turn is a "closer semantic match", as a Java-side set *could* contain `null`, so "fully plumbing" `T?` through `JavaSet<T>` would be "more correct", e.g. Java `HastSet` *can* contain `null` values.  The non-generic `JavaSet` supports null, e.g. it provides `Add(object?)`.  `JavaSet.Add(T)` should be `JavaSet.Add(T?)`.
 5. which just "moves" the `!` usage to the array insertion:
    ```csharp    
    foreach (T? item in this) {
        array [array_index + i++] = item!;
    }
    ```
    which "feels better", yet…is still worrisome.

 7. …which in turn makes me worried about our initial nullability effort. :-(

Reply:
Indeed, using `T?` made the results much cleaner.  Removed most instances of `[return: MaybeNull]`.